### PR TITLE
Package code, make pip installable, and add entrypoints

### LINKED ISSUE #1: Package code and make pip installable
It would be nice to make this code pip installable and to use entrypoints to make a vanity script through which the scripts can be run. I'd be happy to send a PR next week. Have a nice weekend, cheers on your new manuscript.

### PR DESCRIPTION
**Pull request by cthoyt**
Closes #1

This is a really big PR, so very quickly it does a few things:

- Reorganizes repository to follow python community standard
- Adds setup.cfg and setup.py and update imports so it can be pip installed (+ updated README)
- Reorganizes code that lived in `if __name__ == '__main__'` to have enclosing `main()` functions
- Adds a command line interface using entrypoints (+ updated README)

I think there are still a couple things that need to be explained (such as #2) before I can attempt to reproduce any of the results described in the manuscript. Adding more explanation to the README would be greatly appreciated.

I would be happy to explain any of the changes I made in more detail, or give you some resources that I've read along the way while I was learning all of this packaging stuff.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,208 @@
...
+.venv
+pip-selfcheck.json
+### JetBrains template
+# Covers JetBrains IDEs: IntelliJ, RubyMine, PhpStorm, AppCode, PyCharm, CLion, Android Studio and WebStorm
+# Reference: https://intellij-support.jetbrains.com/hc/en-us/articles/206544839
+
+# User-specific stuff
 .idea
-__pycache__
-.ipynb_checkpoints/
```
Comment (Dierme) [inline .gitignore]: Why deleting `__pycache__` and `.ipynb_checkpoints/` ?

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,208 @@
...
+.Trashes
+.VolumeIcon.icns
+.com.apple.timemachine.donotpresent
+
+# Directories potentially created on remote AFP share
+.AppleDB
+.AppleDesktop
+Network Trash Folder
+Temporary Items
+.apdisk
```
Comment (Dierme) [inline .gitignore]: Why overloading .gitignore so much? 90% of that is ignored here will never appear in this type of project. Just add the most essential things to the .gitignore to keep it simple.

File: setup.cfg [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+long_description_content_type = text/markdown
+
+# Search tags
+classifiers =
+    Programming Language :: Python
+    Programming Language :: Python :: 3.7
+    Programming Language :: Python :: 3.6
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3 :: Only
+    Topic :: Scientific/Engineering :: Bio-Informatics
```
Comment (Dierme) [inline setup.cfg]: Not a bioinformatics project. Delete this tag. You can put the tags like in the manuscript or just leave it like that for me to update proper tags later.

File: setup.cfg [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    Topic :: Scientific/Engineering :: Chemistry
+keywords =
+
+[options]
+install_requires =
+    numpy
+    torch
```
Comment (Dierme) [inline setup.cfg:22]: Installing torch fails on Windows. Is it possible to make setuptools to be aware of the environment and run system specific installs?

File: setup.cfg [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+# Random options
+zip_safe = false
+include_package_data = True
+python_requires = >=3.5
```
Comment (Dierme) [inline setup.cfg]: should be 3.6

File: src/latent_gan/autoencoder/autoencoder.py [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
         model_version = DEFAULT_MODEL_VERSION
     path = os.path.join(os.path.dirname(os.path.realpath(__file__)), model_version)
-    model = ddc.DDC(model_name=path)
+    return load_model_from_path(path)
+
 
-    return model
+def load_model_from_path(path):
+    from ddc_pub import ddc_v3 as ddc
+    return ddc.DDC(model_name=path)
```
Comment (Dierme) [inline src/latent_gan/autoencoder/autoencoder.py]: I see the idea that you tried to do to make it more flexible and optimize the imports. However it is not supposed to work this way. Reverse this one for now. I will create a stub for the autoencoder to illustrate how it works later.

File: setup.cfg (hunk 4 above)
Comment (Dierme) [inline setup.cfg:22]: Installing torch fails on windows. Is it possible to make setuptools aware of the environment so it runs system specific installations?

File: setup.cfg (hunk 3 above)
Comment (Dierme) [inline setup.cfg]: Not a bioinformatics project, remove this tag. I will add tags from the manuscript later.

File: setup.cfg (hunk 5 above)
Comment (Dierme) [inline setup.cfg]: should be 3.6

File: src/latent_gan/autoencoder/autoencoder.py [hunk 7]
```diff
@@ -1,14 +1,17 @@
-from ddc_pub import ddc_v3 as ddc
+# -*- coding: utf-8 -*-
+
 import os
 
 DEFAULT_MODEL_VERSION = '9784435'
 
 
 def load_model(model_version=None):
-    # Import model
```
Comment (Dierme) [inline src/latent_gan/autoencoder/autoencoder.py]: I think a comment here is a good thing. But maybe name it _# Loading the autoencoder model_ ?

File: src/latent_gan/decode.py [hunk 8]
```diff
@@ -1,16 +1,21 @@
...
+from latent_gan.autoencoder import autoencoder
 
-def decode(latent_mols_file, output_smiles_file_path=None, message=''):
+
+def decode(model_path, latent_mols_file, output_smiles_file_path=None, message=''):
     print("BEGIN")
     print("Message: %s " % message)
     sys.stdout.flush()
-    model = autoencoder.load_model()
+    model = autoencoder.load_model_from_path(model_path)
```
Comment (Dierme) [inline src/latent_gan/decode.py]: Same here for the function signature and _load_model_from_path_ - nice idea, I will use something like this, but let me insert the stub of the autoencoder first. So, reverse those two changes for now please.

File: src/latent_gan/decode_paraller.py [hunk 9]
```diff
@@ -32,10 +37,10 @@ def latent_to_smile_string(latent, model, file_path):
             smiles_file.flush()
 
 
-def decode(latent_mols_file, output_smiles_file_path=None):
+def decode(model_path, latent_mols_file, output_smiles_file_path=None):
     print("BEGIN")
     sys.stdout.flush()
-    model = autoencoder.load_model()
+    model = autoencoder.load_model_from_path(model_path)
```
Comment (Dierme) [inline src/latent_gan/decode_paraller.py]: same as above

File: src/latent_gan/encode.py [hunk 10]
```diff
@@ -1,18 +1,22 @@
...
+def encode(model_path, smiles_file, output_smiles_file_path=None):
+    """Encoding smile strings into latent vectors that are used to train GAN.
+
+    :param model_path:
     :param smiles_file:
     :param output_smiles_file_path:
     :return:
     """
-    model = autoencoder.load_model()
+    model = autoencoder.load_model_from_path(model_path)
```
Comment (Dierme) [inline src/latent_gan/encode.py]: same

File: src/latent_gan/sample.py [hunk 11]
```diff
@@ -55,7 +62,11 @@ def sample(generator_path, output_sampled_latent_file, sample_number=50000, mess
 
     args = {k: v for k, v in vars(parser.parse_args()).items() if v is not None}
 
-    if args['decode_sampled'] == True:
-        if 'output_decoded_smiles_file' not in args:
-            raise Exception('Argument \'output_decoded_smiles_file\' should be set if \'decode\'=True')
+    if args['decode_sampled'] is True and 'output_decoded_smiles_file' not in args:
+        raise Exception('Argument \'output_decoded_smiles_file\' should be set if \'decode\'=True')
+
```
Comment (Dierme) [inline src/latent_gan/sample.py:67]: Nice catch here! Indeed, when comparing booleans _is_ works faster.

**Review by Dierme — CHANGES_REQUESTED** [review]
Thank you a lot for the pull request!

A lot of good things and cleaning and introducing the setup tools. I have left a few questions. Mainly, I want a simplified .gitignore with only essentials in it and that you reverse changes that you made to the autoencoder. 

It would also be great if you can make setuptools to be aware of the environment and run system specific command. As of now it fails in Windows environment. That is not a requirement of course  and can be thought about in the future versions.

Great work overall! Let me know if you will do the changes. Have a good day!

File: .gitignore (hunk 1 above)
Comment (cthoyt) [reply .gitignore]: They just got moved, but this is superseded by your request to keep it minimal. Will fix in a minute!

File: setup.cfg (hunk 4 above)
Comment (cthoyt) [reply setup.cfg:22]: I think PyTorch has some complicated instructions for installing on Windows. As the Python community moves towards declarative setups (like this one) there is less room to do checking for that sort of stuff. I think the best solution would be to add something to the README for windows users to install PyTorch with something like Anaconda first.

Comment (cthoyt) [reply setup.cfg:22]: Updated with 582f6f0

File: src/latent_gan/autoencoder/autoencoder.py (hunk 7 above)
Comment (cthoyt) [reply src/latent_gan/autoencoder/autoencoder.py]: That's a good point, maybe as a docstring would be better.

File: src/latent_gan/decode.py (hunk 8 above)
Comment (cthoyt) [reply src/latent_gan/decode.py]: Okay, I'm sorry I made some assumptions about how this works. I'm looking forwards to seeing your example! For now, I've reverted this code to look the same as before (except I moved around where the imports were just so I didn't get any `NameError`s.)

**Comment by cthoyt** [comment]
Hi @Dierme I think I've addressed all of your changes. In its current state, this PR makes the repository installable and accomplishes my original goals. The next thing I'm looking forwards to is your example autoencoder so I can try actually running the code myself!

**Pull request closed without merging**